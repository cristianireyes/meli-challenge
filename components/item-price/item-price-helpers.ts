export const parsedAmount = (amount: number): string => {
  const formatter = new Intl.NumberFormat('es-AR');

  return formatter.format(amount);
};

export const CurrencyMap: Record<string, string> = {
  ARS: '$',
  BRL: 'R$',
  USD: 'U$S',
};
