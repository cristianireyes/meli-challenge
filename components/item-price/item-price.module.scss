@mixin base {
  display: flex;
  gap: 0.25rem;
  position: relative;
  width: fit-content;

  .decimals {
    position: absolute;
    right: -1rem;
    font-size: 0.75rem;
  }
}

.price {
  @include base;
  font-size: 1.25rem;
}

.priceLarge {
  @include base;
  font-size: 2rem;

  .decimals {
    font-size: 0.875rem;
  }
}