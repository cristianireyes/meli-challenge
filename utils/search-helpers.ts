import { Item } from './types';

export const getCategories = (filters: any[]): string[] => {
  const categoriesFiltered = filters.find(filter => filter.id === 'category');

  const categories = categoriesFiltered?.values[0]?.path_from_root?.map(
    (category: any) => category.name,
  );

  return categories ?? [];
};

const getPictureURL = (id: string): string => {
  return `https://http2.mlstatic.com/D_NQ_NP_${id}-V.webp`;
};

export const getParsedItem = (item: any): Item => {
  const [amount, decimals] = item.price.toString().split('.');

  return {
    id: item.id,
    condition: item.condition,
    freeShipping: item.shipping.free_shipping,
    picture: getPictureURL(item.thumbnail_id),
    price: {
      amount: Number(amount),
      currency: item.currency_id,
      ...(decimals && { decimals: Number(decimals) }),
    },
    title: item.title,
    soldQuantity: item.sold_quantity,
    address: {
      stateName: item.seller_address.state.name,
    },
  };
};

/**
 * get the first four elements
 * @param data data provided by Meli
 * @returns
 */
export const getItems = (data: any): Item[] => {
  const items = data.slice(0, 4);

  return items.map((item: any): Item => getParsedItem(item));
};
