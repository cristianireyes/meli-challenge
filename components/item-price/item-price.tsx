import { FunctionComponent } from 'react';
import { Price } from 'utils/types';
import { CurrencyMap, parsedAmount } from './item-price-helpers';
import styles from './item-price.module.scss';

type Variant = 'normal' | 'large';
interface ItemPriceProps {
  price: Price;
  variant?: Variant;
}

const styleVariantMap: Record<Variant, string> = {
  normal: styles.price,
  large: styles.priceLarge,
};

export const ItemPrice: FunctionComponent<ItemPriceProps> = ({
  price,
  variant = 'normal',
}) => (
  <div className={styleVariantMap[variant]}>
    <span>{CurrencyMap[price.currency]}</span>
    <span>{parsedAmount(price.amount)}</span>
    <span className={styles.decimals}>{price.decimals}</span>
  </div>
);
