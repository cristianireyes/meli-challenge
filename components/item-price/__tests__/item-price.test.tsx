import { render } from '@testing-library/react';
import { Price } from 'utils/types';
import { ItemPrice } from '../item-price';

/**
 * Mocks
 */
const price: Price = {
  amount: 20000,
  currency: 'USD',
  decimals: 20,
};

/**
 * Tests
 */

describe('item-price component', () => {
  it('should show the price of the item with decimals', () => {
    const { container, getByText } = render(<ItemPrice price={price} />);

    expect(getByText('20.000')).toBeDefined();
    expect(getByText('U$S')).toBeDefined();
    expect(getByText(`${price.decimals}`)).toBeDefined();
    expect(container).toMatchSnapshot();
  });

  it('should show the price of the item without decimals', async () => {
    const priceWithoutDecimals: Price = {
      ...price,
      decimals: undefined,
    };
    const { container, getByText, queryByText } = render(
      <ItemPrice price={priceWithoutDecimals} />,
    );

    expect(getByText('20.000')).toBeDefined();
    expect(getByText('U$S')).toBeDefined();
    expect(queryByText(`${priceWithoutDecimals.decimals}`)).toBeNull();
    expect(container).toMatchSnapshot();
  });

  it('should show the price of the item with Argentine currency', async () => {
    const priceWithArgentineCurrency: Price = {
      ...price,
      currency: 'ARS',
    };
    const { container, getByText, queryByText } = render(
      <ItemPrice price={priceWithArgentineCurrency} />,
    );

    expect(getByText('20.000')).toBeDefined();
    expect(getByText('$')).toBeDefined();
    expect(queryByText(`${priceWithArgentineCurrency.decimals}`)).toBeDefined();
    expect(container).toMatchSnapshot();
  });
});
