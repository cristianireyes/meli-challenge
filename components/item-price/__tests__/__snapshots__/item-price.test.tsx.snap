// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`item-price component should show the price of the item with Argentine currency 1`] = `
<div>
  <div
    class="price"
  >
    <span>
      $
    </span>
    <span>
      20.000
    </span>
    <span
      class="decimals"
    >
      20
    </span>
  </div>
</div>
`;

exports[`item-price component should show the price of the item with decimals 1`] = `
<div>
  <div
    class="price"
  >
    <span>
      U$S
    </span>
    <span>
      20.000
    </span>
    <span
      class="decimals"
    >
      20
    </span>
  </div>
</div>
`;

exports[`item-price component should show the price of the item without decimals 1`] = `
<div>
  <div
    class="price"
  >
    <span>
      U$S
    </span>
    <span>
      20.000
    </span>
  </div>
</div>
`;
